packages/mds-utils/state-machine.ts

```diff
@@ -1,58 +1,126 @@
+    [VEHICLE_EVENTS.missing]: [VEHICLE_STATES.unknown]
   },
   [VEHICLE_STATES.removed]: {
-    [VEHICLE_EVENTS.trip_enter]: VEHICLE_STATES.trip,
-    [VEHICLE_EVENTS.provider_drop_off]: VEHICLE_STATES.available,
-    [VEHICLE_EVENTS.deregister]: VEHICLE_STATES.inactive
+    [VEHICLE_EVENTS.agency_drop_off]: [VEHICLE_STATES.available],
+    [VEHICLE_EVENTS.decommissioned]: [VEHICLE_STATES.removed],
+    [VEHICLE_EVENTS.provider_drop_off]: [VEHICLE_STATES.available],
+    [VEHICLE_EVENTS.unspecified]: [VEHICLE_STATES.available]
   },
   [VEHICLE_STATES.reserved]: {
-    [VEHICLE_EVENTS.trip_start]: VEHICLE_STATES.trip,
-    [VEHICLE_EVENTS.cancel_reservation]: VEHICLE_STATES.available
-  },
-  [VEHICLE_STATES.trip]: {
-    [VEHICLE_EVENTS.trip_leave]: VEHICLE_STATES.elsewhere,
-    [VEHICLE_EVENTS.trip_end]: VEHICLE_STATES.available
+    [VEHICLE_EVENTS.comms_lost]: [VEHICLE_STATES.unknown],
+    [VEHICLE_EVENTS.missing]: [VEHICLE_STATES.unknown],
+    [VEHICLE_EVENTS.reservation_cancel]: [VEHICLE_STATES.available],
+    [VEHICLE_EVENTS.trip_start]: [VEHICLE_STATES.on_trip],
+    [VEHICLE_EVENTS.unspecified]: [VEHICLE_STATES.available]
   },
-  [VEHICLE_STATES.unavailable]: {
-    [VEHICLE_EVENTS.service_start]: VEHICLE_STATES.available,
-    [VEHICLE_EVENTS.deregister]: VEHICLE_STATES.inactive,
-    [VEHICLE_EVENTS.agency_pick_up]: VEHICLE_STATES.removed,
-    [VEHICLE_EVENTS.provider_pick_up]: VEHICLE_STATES.removed
+  [VEHICLE_STATES.unknown]: {
+    [VEHICLE_EVENTS.agency_drop_off]: [VEHICLE_STATES.available],
+    [VEHICLE_EVENTS.agency_pick_up]: [VEHICLE_STATES.removed],
+    [VEHICLE_EVENTS.comms_restored]: [
+      VEHICLE_STATES.available,
+      VEHICLE_STATES.elsewhere,
+      VEHICLE_STATES.reserved,
+      VEHICLE_STATES.on_trip,
+      VEHICLE_STATES.non_operational
+    ],
+    [VEHICLE_EVENTS.decommissioned]: [VEHICLE_STATES.removed],
+    [VEHICLE_EVENTS.provider_drop_off]: [VEHICLE_STATES.available],
+    [VEHICLE_EVENTS.unspecified]: [VEHICLE_STATES.available, VEHICLE_STATES.removed]
   }
 }
 
-const getNextState = (currStatus: VEHICLE_STATE, nextEvent: VEHICLE_EVENT): VEHICLE_STATE | undefined => {
+const getNextStates = (currStatus: VEHICLE_STATE, nextEvent: VEHICLE_EVENT): Array<VEHICLE_STATE> | undefined => {
   return stateTransitionDict[currStatus]?.[nextEvent]
 }
 
+function isStateTransitionValid(eventA: VehicleEvent, eventB: VehicleEvent) {
+  const currStates = EVENT_STATES_MAP[eventA.event_type]
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const allNextStates: any = currStates.reduce((acc, currState) => {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    // See if it's possible to transition to any states using eventB's event_type
+    const nextStates: any = getNextStates(currState, eventB.event_type)
+    if (nextStates) {
+      return acc.concat(nextStates)
+    }
+    return acc
+  }, [])
+  return allNextStates.length > 0
+}
```

Comment:
Also probably want to only do the looping for every event_type except the last one, because we know the resultant state from the last event_type as its explicitly encoded in the payload.